### PR TITLE
salt: 3004.1 - bugfixes

### PR DESCRIPTION
**Pull request by takeda**
###### Description of changes

Contains following fixes:

- crashes with Jinja2 3.1.0 (in nixpkgs)
- crashes on Python 3.10+
- TCP transport broken
- doesn't build on Darwin
- doesn't work with pyzmq 22.3.0 (in nixpkgs)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by takeda** [comment]
Result of `nixpkgs-review pr 172129` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>pepper</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>salt</li>
  </ul>
</details>

File: pkgs/tools/admin/salt/default.nix [hunk 1]
```diff
@@ -39,18 +25,46 @@ py.pkgs.buildPythonApplication rec {
...
+    substituteInPlace "salt/transport/tcp.py" \
+      --replace 'payload = self.pack_publish(package)' 'package = self.pack_publish(package)'
+
+    # 3004.1: requirement of pyzmq was restricted to <22.0.0; looks like that req was incorrect
+    # https://github.com/saltstack/salt/commit/070597e525bb7d56ffadede1aede325dfb1b73a4
+    # https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=259279
+    # https://github.com/saltstack/salt/pull/61163
+    substituteInPlace requirements/zeromq.txt \
+      --replace 'pyzmq>=17.0.0,<22.0.0 ; python_version < "3.9"' 'pyzmq>=17.0.0 ; python_version < "3.9"'
+    substituteInPlace requirements/zeromq.txt \
```
Comment (SuperSandro2000) [inline pkgs/tools/admin/salt/default.nix:61]:
```suggestion
      --replace 'pyzmq>=17.0.0,<22.0.0 ; python_version < "3.9"' 'pyzmq>=17.0.0 ; python_version < "3.9"' \
```

Comment (takeda) [reply pkgs/tools/admin/salt/default.nix]: @SuperSandro2000 updated

**Comment by takeda** [comment]
Added last minute patch to remove `pyzmq<=20.0.0 ; python_version < "3.6"` (nixpkgs doesn't have python 3.0-3.6 anyway) it looks like there's another bug in salt around treating versions like decimals and 3.10 is considered lower than 3.6.

**Review by Flakebi — APPROVED** [review]
Looks good to me, thanks!

Result of `nixpkgs-review pr 172129` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>pepper</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>salt</li>
  </ul>
</details>

File: maintainers/maintainer-list.nix [hunk 2]
```diff
@@ -12313,6 +12313,12 @@
     githubId = 18656090;
     name = "Yuki Takagi";
   };
+  takeda = {
+    email = "d@kulinski.us";
+    github = "takeda";
+    githubId = 411978;
+    name = "Derek Kulinski";
+  };
```
Comment (mweinelt) [inline maintainers/maintainer-list.nix:12321]: unused and should go into a separate commit either way.

**Comment by takeda** [comment]
@mweinelt @SuperSandro2000 @Flakebi I fixed all recommendations, is there anything else I might need to do?

**Review by Flakebi — APPROVED** [review]
I’m still fine with it

**Comment by mweinelt** [comment]
Do we need to carry that patch downstream? Did you try to get it merged the upstream?

**Comment by takeda** [comment]
@mweinelt all the patches (except TCP transport bug one as they still talk about it) are upstream, but in a version that's not released.

Salt has a really bad release process and every version fixes bugs and introduced bunch of new ones. The TCP transport bug for example was introduced in CVE fix version. It's ridiculous.

**Comment by mweinelt** [comment]
And we can't use `fetchpatch` to pull them?

**Comment by takeda** [comment]
@mweinelt I'm afraid, because in fixes they often make some refactoring etc. But if you want I could use the Jinja2 and zeromq fix from upstream.

**Comment by mweinelt** [comment]
If that is possible it would be preferred.

**Comment by Flakebi** [comment]
I tried pulling the jinja2 fix from upstream but it didn’t really work out (too much changed in-between and the upstream patch is a load bigger than what @takeda wrote).

**Comment by takeda** [comment]
@Flakebi now as you said it I actually remember doing that too at first and the patch was rejected. I thought I was using `fetchpatch` wrong or something.

**Comment by viq** [comment]
https://github.com/saltstack/salt-enhancement-proposals/pull/65 may be of interest to you
